### ISSUE TITLE
bad result from lfw dataset

### ISSUE THREAD

**Issue by ThiagoMateo** [peteryuX/arcface-tf2#8]
![Alicia_Silverstone_0001](https://user-images.githubusercontent.com/48821989/77631147-c3d06b80-6f7e-11ea-8c64-404466293fb5.png)
![Alicia_Silverstone_0002](https://user-images.githubusercontent.com/48821989/77631160-c5019880-6f7e-11ea-8d64-fa44010ec104.png)

hello @peteryuX  i try to test accuracy on 2 images of 1 person in lfw dataset.
The distance is big.

my test code:
```
from absl import app, flags, logging
from absl.flags import FLAGS
import cv2
import os
import numpy as np
import tensorflow as tf

from modules.evaluations import get_val_data, perform_val
from modules.models import ArcFaceModel
from modules.utils import set_memory_growth, load_yaml, l2_norm
from scipy.spatial.distance import cosine


flags.DEFINE_string('cfg_path', './configs/arc_res50.yaml', 'config file path')
flags.DEFINE_string('gpu', '0', 'which gpu to use')
flags.DEFINE_string('img_path', '', 'path to input image')


def main(_argv):
    os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
    os.environ['CUDA_VISIBLE_DEVICES'] = FLAGS.gpu

    logger = tf.get_logger()
    logger.disabled = True
    logger.setLevel(logging.FATAL)
    set_memory_growth()

    cfg = load_yaml(FLAGS.cfg_path)

    model = ArcFaceModel(size=cfg['input_size'],
                         backbone_type=cfg['backbone_type'],
                         training=False)

    ckpt_path = tf.train.latest_checkpoint('./checkpoints/')
    if ckpt_path is not None:
        print("[*] load ckpt from {}".format(ckpt_path))
        model.load_weights(ckpt_path)
    else:
        print("[*] Cannot find ckpt from {}.".format(ckpt_path))
        exit()
    image_fol = "./tmp"
    paths = os.listdir(image_fol)
    embeds = []
    images = []
    flip_images = []
    for path in paths:
        print(path)
        img = cv2.imread(os.path.join(image_fol, path))
        img = cv2.resize(img, (cfg['input_size'], cfg['input_size']))
        img = img.astype(np.float32) / 255.
        # if len(img.shape) == 3:
        #     img = np.expand_dims(img, 0)
        # embeds.append(l2_norm(model(img)))
        images.append(img)
    images = np.array(images)
    def hflip_batch(imgs):
        assert len(imgs.shape) == 4
        return imgs[:, :, ::-1, :]

    flip_images = hflip_batch(images)
    embeds = model(images) + model(flip_images)
    embeds = l2_norm(embeds)

    dist = np.sum(np.square(embeds[0]-embeds[1]))

    print("dist: ", dist)
    # diff = np.subtract([embeds[0]], [embeds[1]])
    # dist = np.sum(np.square(diff), 1)
    # print("diff: ", diff)
    acc = 1 - cosine(embeds[0], embeds[1])
    print("acc: ", acc)

```

the result is:
```
dist:  1.1708782
acc:  0.4145609736442566
```

How can i achive 99.35% accuracy on lfw?

**Comment by peteryuX** [comment]
Hi @ThiagoMateo, nice to hear from you!

I know you use the resnet50+CCrop version (I get the same distance from the paired image).

However, the **99.35%** acc is based on the threshold **1.32** which can found from the result produced by running test.py like bellow:
```
[*] Loading LFW, AgeDB30 and CFP-FP...
[*] Perform Evaluation on LFW...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 94/94 [00:16<00:00,  5.73it/s]
    acc 0.9935, th: 1.32
```
So the distance 1.1708782 is bellow the threshold and it means the same person for this situation.

**Comment by HoangTienDuc** [comment]
hello @peteryuX . Thanks to your great code. 
I try to evaluate your [resnet50 model](https://drive.google.com/file/d/1HasWQb86s4xSYy36YbmhRELg9LBmvhvt/view?usp=sharing) without ccrop by using a [subset of lfw](https://drive.google.com/drive/folders/1g2GylgE_cuAggKs9DmvB6cA-8sTIQvZo) dataset.
 
This subset was cropped and was aligned by MTCNN. 
```
I0409 22:16:14.182628 139845503842112 utils.py:18] Detect 1 Physical GPUs, 1 Logical GPUs.
[*] load ckpt from ./checkpoints/e_8_b_40000.ckpt
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:01<00:00,  1.34it/s]
    acc 0.8375, th: 1.38
```
it seem that the result is very low compared with testing on your lfw binary data

My code is below:
```
def arc_preprocess(img, image_shape):
    img = cv2.resize(img, (image_shape[0], image_shape[1]))
    img = img.astype(np.float32) / 255.
    return img
```
in main function:
```
src_pair = "/home/tienduchoang/Videos/face/dataset/tmp.txt"
src_img_dir = "/home/tienduchoang/Videos/face/dataset/tmp"
image_shape = (112, 112, 3)

pairs = read_pairs(src_pair)

path_list, issame_list = get_paths(src_img_dir, pairs)
        
images = []
for path in path_list:
    image = cv2.imread(path)
    image = arc_preprocess(image, image_shape)
    images.append(image)
images = np.array(images)
acc_lfw, best_th = perform_val(
    cfg['embd_shape'], cfg['batch_size'], model, images, issame_list,
    is_ccrop=cfg['is_ccrop'])
print("    acc {:.4f}, th: {:.2f}".format(acc_lfw, best_th))
```

**Comment by peteryuX** [comment]
Hi @HoangTienDuc! Thanks for asking.
Did you ever modify the code bellow
https://github.com/peteryuX/arcface-tf2/blob/6da8812e0b65772c0d2b83f896e291d547373900/modules/evaluations.py#L118
I hope you didn't do prepossessing again, which would be a problem. Let me know more detail for better understanding of your situation to help you. =)

**Comment by HoangTienDuc** [comment]
hi @peteryuX  I deleted the code that you mentioned.
i test a [subset](https://drive.google.com/drive/folders/1g2GylgE_cuAggKs9DmvB6cA-8sTIQvZo) of lfw. this subset contain several folders containing people images and a file pairs.txt.
i have some small change to your great code to test on this subset. i check the accuray on your arcface resnet50 model. the acurracy is 0.8375.
If you have time, you can take a look at the code I have [modified](https://drive.google.com/drive/folders/1g2GylgE_cuAggKs9DmvB6cA-8sTIQvZo).

**Comment by peteryuX** [comment]
Hi @HoangTienDuc ! I tried your code and got same results. I cannot find any mistakes in the code right now... That's really interesting... Did you try more paired samples from lfw dataset? If you find any  news, please let me know. I feel confused too.

**Comment by HoangTienDuc** [comment]
hi @peteryuX  Thanks for your enthusiasm
This [subset](https://drive.google.com/drive/folders/1g2GylgE_cuAggKs9DmvB6cA-8sTIQvZo) was aligned base on facenet [alignment](https://github.com/davidsandberg/facenet/wiki/Validate-on-lfw) method.

- I tried [another](https://github.com/ZhaoJ9014/face.evoLVe.PyTorch/tree/master/align) alignment method that you mentioned in this [issue](https://github.com/peteryuX/arcface-tf2/issues/6).
the result after alignment i put in [here](https://drive.google.com/drive/folders/1bjR9MdYRG84zVhqRes1wkJlRV8oFtX_d)

- i also convert color (BGR -> RGB) after reading image. 


Finally, the accuracy is ~97%

**Comment by peteryuX** [comment]
@HoangTienDuc Thank you！How fast you are. You are right, the opencv is bgr format, which is not match with the tf training version. The aliment things is totally fresh for me, let me learn more, thank you very much！

**Comment by HoangTienDuc** [comment]
Hi @peteryuX your work is awesome. Author can not cover all thing. It is common thing. Thank you very much.

**Comment by FelixZhang7** [comment]
Hello@HoangTienDuc，may i ask why      acc = 1 - cosine(embeds[0], embeds[1])?

**Comment by HoangTienDuc** [comment]
hello @FelixZhang7  It is exactly similarity, not acc.
